cmd/juju/commands/upgrademodel.go

```diff
@@ -396,14 +403,328 @@ func (c *upgradeJujuCommand) Run(ctx *cmd.Context) (err error) {
+	if err != nil {
+		return err
+	}
+	defer controllerClient.Close()
+
+	modelConfigClient, err := c.getModelConfigAPI()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer modelConfigClient.Close()
+
+	attrs, err := modelConfigClient.ModelGet()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cfg, err := config.New(config.NoDefaults, attrs)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	agentVersion, ok := cfg.AgentVersion()
+	if !ok {
+		// Can't happen. In theory.
+		return errors.New("incomplete model configuration")
+	}
+	if majorDiff := jujuversion.Current.Major - agentVersion.Major; majorDiff != 0 && majorDiff != 1 {
+		// This version of juju client cannot upgrade the running
+		// model version (can't guarantee API compatibility).
+		return errors.Errorf("cannot upgrade a %s model with a %s client", agentVersion, jujuversion.Current)
+	}
+
+	if c.Version == agentVersion {
+		return errUpToDate
+	}
+
+	controllerModelConfig, err := controllerClient.ModelConfig()
+	if err != nil && !params.IsCodeUnauthorized(err) {
+		return err
+	}
+	haveControllerModelPermission := err == nil
+	isControllerModel := haveControllerModelPermission && cfg.UUID() == controllerModelConfig[config.UUIDKey]
+
+	if c.BuildAgent {
+		// For UploadTools, model must be the "controller" model,
+		// that is, modelUUID == controllerUUID
+		if !haveControllerModelPermission {
+			return errors.New("--build-agent can only be used with the controller model but you don't have permission to access that model")
+		}
+		if !isControllerModel {
+			return errors.Errorf("--build-agent can only be used with the controller model")
+		}
+		if targetVersion != version.Zero {
+			return errors.Errorf("--build-agent cannot be used with --agent-version together")
+		}
+	}
+
+	// Decide the target version to upgrade.
+	if targetVersion != version.Zero {
+		if targetVersion, err = c.upgradeWithTargetVersion(
+			ctx, modelUpgrader, isControllerModel, c.DryRun,
+			modelType, targetVersion, agentVersion,
+		); err != nil {
+			return err
+		}
+		return nil
+	} else if c.BuildAgent {
+		if targetVersion, err = uploadTools(modelUpgrader, c.BuildAgent, agentVersion, c.DryRun); err != nil {
+			return block.ProcessBlockedError(err, block.BlockChange)
+		}
+		builtMsg := " (built from source)"
+		fmt.Fprintf(ctx.Stdout,
+			"no prepackaged agent binaries available, using local agent binary %v%s\n",
+			targetVersion, builtMsg,
+		)
+		targetVersion, err = c.notifyControllerUpgrade(ctx, modelUpgrader, targetVersion, c.DryRun)
+		if err != nil {
+			return err
+		}
+		return nil
+	} else {
+		targetVersion, err = c.notifyControllerUpgrade(
+			ctx, modelUpgrader,
+			version.Zero, // no target version provided, we figure it out on the server side.
+			c.DryRun,
+		)
+		if err == nil {
+			// All good!
+			// Upgraded to a next stable version or the newest stable version.
+			logger.Debugf("upgraded to a next version or latest stable version")
+			return nil
+		}
```

Comment:
this can be consolidated to avoid the duplication of notifyControllerUpgrade
just have the buildagent branch set the target version instead of version.Zero